wasm_exec/internal/spfunc/spfunc.go

```diff
@@ -0,0 +1,186 @@
+package spfunc
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/internal/leb128"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+const debugMode = false
+
+func MustCallFromSP(expectSP bool, proxied *wasm.HostFunc) *wasm.ProxyFunc {
+	if ret, err := callFromSP(expectSP, proxied); err != nil {
+		panic(err)
+	} else {
+		return ret
+	}
+}
+
+// callFromSP generates code to call a function with the provided signature.
+// The returned function has a single api.ValueTypeI32 parameter of SP. Each
+// parameter is read at 8 byte offsets after that, and each result is written
+// at 8 byte offsets after the parameters.
+//
+// # Parameters
+//
+//   - expectSP: true if a constructor or method invocation. The last result is
+//     an updated SP value (i32), which affects the result memory offsets.
+func callFromSP(expectSP bool, proxied *wasm.HostFunc) (*wasm.ProxyFunc, error) {
+	params := proxied.ParamTypes
+	results := proxied.ResultTypes
+	if (8+len(params)+len(results))*8 > 255 {
+		return nil, errors.New("TODO: memory offset larger than one byte")
+	}
+
+	if debugMode {
+		fmt.Printf("(func $%s.proxy (param $%s %s)", proxied.Name, "sp", wasm.ValueTypeName(wasm.ValueTypeI32))
+	}
+
+	var localTypes []api.ValueType
+
+	resultSpOffset := 8 + len(params)*8
+	resultSPIndex := byte(0)
+	if len(results) > 0 {
+		if debugMode {
+			fmt.Printf(" (local %s %s)", wasm.ValueTypeName(wasm.ValueTypeI32), wasm.ValueTypeName(wasm.ValueTypeI64))
+		}
+		localTypes = append(localTypes, api.ValueTypeI32, api.ValueTypeI64)
+		if expectSP {
+			if debugMode {
+				fmt.Printf(" (local %s)", wasm.ValueTypeName(wasm.ValueTypeI32))
+			}
+			resultSPIndex = 3
+			resultSpOffset += 8
+			localTypes = append(localTypes, api.ValueTypeI32)
+		}
+	}
+
+	// Load all parameters onto the stack.
+	var code []byte
+	for i, t := range params {
+		if debugMode {
+			fmt.Printf("\n;; param[%d]=%s\n", i, wasm.ValueTypeName(t))
+		}
+
+		// First, add the memory offset to load onto the stack.
+		offset := 8 + int(i*8)
+		code = compileAddOffsetToSP(code, 0, offset)
+
+		// Next, load stack parameter $i from memory at that offset.
+		switch t {
+		case api.ValueTypeI32:
+			if debugMode {
+				fmt.Println(wasm.OpcodeI32LoadName)
+			}
+			code = append(code, wasm.OpcodeI32Load, 0x2, 0x0) // alignment=2 (natural alignment) staticOffset=0
+		case api.ValueTypeI64:
+			if debugMode {
+				fmt.Println(wasm.OpcodeI64LoadName)
+			}
+			code = append(code, wasm.OpcodeI64Load, 0x3, 0x0) // alignment=3 (natural alignment) staticOffset=0
+		default:
+			panic(errors.New("TODO: param " + api.ValueTypeName(t)))
+		}
+	}
+
+	// Now that all parameters are on the stack, call the function
+	callFuncPos := len(code) + 1
+	if debugMode {
+		fmt.Printf("\n%s 0\n", wasm.OpcodeCallName)
+	}
+	code = append(code, wasm.OpcodeCall, 0)
```

Comment:
I think I want comment on this - zero is a place holder and will be replaced later by the actual index